core/src/main/scala/org/apache/spark/network/netty/SparkTransportConf.scala

```diff
@@ -36,16 +36,30 @@ object SparkTransportConf {
    * @param numUsableCores if nonzero, this will restrict the server and client threads to only
    *                       use the given number of cores, rather than all of the machine's cores.
    *                       This restriction will only occur if these properties are not already set.
+   * @param role           optional role, could be driver, executor, worker and master. Default is
+   *                      [[None]], means no role specific configurations.
    */
-  def fromSparkConf(_conf: SparkConf, module: String, numUsableCores: Int = 0): TransportConf = {
+  def fromSparkConf(
+      _conf: SparkConf,
+      module: String,
+      numUsableCores: Int = 0,
+      role: Option[String] = None): TransportConf = {
     val conf = _conf.clone
-
-    // Specify thread configuration based on our JVM's allocation of cores (rather than necessarily
-    // assuming we have all the machine's cores).
-    // NB: Only set if serverThreads/clientThreads not already set.
+    // specify default thread configuration based on our JVM's allocation of cores (rather than
+    // necessarily assuming we have all the machine's cores).
     val numThreads = NettyUtils.defaultNumThreads(numUsableCores)
-    conf.setIfMissing(s"spark.$module.io.serverThreads", numThreads.toString)
-    conf.setIfMissing(s"spark.$module.io.clientThreads", numThreads.toString)
+    // module threads configuration
+    val (modServerThreads, modClientThreads) =
```

Comment:
I think this is a little hard to follow. Something like this seems better:

```
Seq("serverThreads", "clientThreads").foreach { suffix =>
  val value = role.flatMap { r => conf.getOption(s"spark.$r.$module.io.$suffix") }.getOrElse(numThreads)
  conf.setIfMissing(s"spark.$module.io.$suffix", value.toString)
}
```

Didn't check whether that works. Which is why you should add unit tests.

Reply:
ok, let me try

Reply:
@vanzin , changed code by following your style. My code has some difference than yours because I need to override module's config with role's if any. the order is role > module > default. It's reflected in the unit test.